File: solution/009PalindromeNumber/009PalindromeNumber.go
package main

import "fmt"

func isPalindrome(x int) bool {
	if x < 0 {
		return false
	} else if x < 10 {
		return true
	}
	n := 1
	y := x
	for y != 0 {
		y /= 10
		n *= 10
	}
	n /= 10
	for x != 0 {
		if x/n != x-(x/10)*10 {
			return false
		}
		x = x - (x/n)*n
		x = x / 10
		n /= 100
	}
	return true
}

func main() {
	fmt.Printf("%v\n", isPalindrome(-121))
	fmt.Printf("%v\n", isPalindrome(0))
	fmt.Printf("%v\n", isPalindrome(1234))
	fmt.Printf("%v\n", isPalindrome(1221))
	fmt.Printf("%v\n", isPalindrome(121))
	fmt.Printf("%v\n", isPalindrome(12))
	fmt.Printf("%v\n", isPalindrome(-2147447412))
}

File: solution/009PalindromeNumber/009PalindromeNumber.rs
impl Solution {
    pub fn is_palindrome(x: i32) -> bool {
        if x < 0 {
            return false;
        }
        let mut n = 0;
        let mut y = x;
        let mut d = 0;
        while y != 0 {
            n += 1;
            y /= 10;
            d = if d == 0 {1} else {d*10};
        }
        y = x;
        while n > 1 && y != 0 {
            if y % 10 != y / d {
                return false;
            }
            y -= (y/d)*d;
            y /= 10;
            n -= 2;
            d /= 100;
        }
        true
    }
}
